File: packages/globstar/src/matcher/parser.js
// Recursive-descent glob parser. Bytes → AST.
// Implements GLOB_SPEC.md §3 grammar with the byte-level conventions of §2.

import {
  N_SEPARATOR,
  lit,
  sep,
  anyChar,
  star,
  globstar,
  klass,
  brace,
  concat,
  classItemByte,
  classItemRange,
  nodeToLiteralBytes,
} from "./ast.js";
import { GlobError, MAX_PATTERN_LEN, MAX_BRACE_NESTING } from "./error.js";
import { toBytes } from "./utf8.js";

// Byte literals for branch tests in the hot scan loop.
const BACKSLASH = 0x5c;
const SLASH = 0x2f;
const QUESTION = 0x3f;
const STAR = 0x2a;
const LBRACK = 0x5b;
const RBRACK = 0x5d;
const LBRACE = 0x7b;
const RBRACE = 0x7d;
const COMMA = 0x2c;
const BANG = 0x21;
const CARET = 0x5e;
const DASH = 0x2d;

// Sequence context: whether we're inside a brace, plus — for the
// globstar segment-ownership test (§8.1) — the brace's *expanded-form*
// neighbors: `{A,B}` is the union of the patterns with the brace
// replaced by each branch (§7), so a `**` at a branch edge is judged
// against what sits outside the brace.
const CTX_TOP = Object.freeze({ brace: false, prevBoundary: true, nextBoundary: true });

// Is the point before the next atom a segment boundary in the expanded
// form? Start-of-pattern and after-`/` are boundaries; a branch start
// inherits from outside the `{` — so `a{**,x}b` degrades its `**` while
// `{**,x}/b` keeps a real globstar. Judged on parsed `nodes` rather
// than raw bytes so escape sequences are handled correctly.
function boundaryBefore(nodes, ctx) {
  if (nodes.length === 0) return ctx.prevBoundary;
  return nodes[nodes.length - 1].tag === N_SEPARATOR;
}

// Mirror of `boundaryBefore` for the byte after an atom: pattern end
// (`undefined`) and `/` are boundaries; a branch end (`,` / `}`)
// inherits from outside the `}`. `undefined` also covers unterminated
// braces, whose errors surface later — the value is then a don't-care.
function boundaryAfter(next, ctx) {
  if (next === undefined || next === SLASH) return true;
  if (next === COMMA || next === RBRACE) return ctx.brace && ctx.nextBoundary;
  return false;
}

export function parse(input) {
  const bytes = toBytes(input);
  if (bytes.length === 0) throw new GlobError("Empty");
  if (bytes.length > MAX_PATTERN_LEN) {
    throw new GlobError("TooLong", { len: bytes.length, max: MAX_PATTERN_LEN });
  }

  const state = { input: bytes, pos: 0, brace_depth: 0, maybe_sep_distribution: false };

  // Leading `!` flips the result on each. Parity decides final negation.
  let negationCount = 0;
  while (state.pos < bytes.length && bytes[state.pos] === BANG) {
    negationCount++;
    state.pos++;
  }

  const body = parseSequence(state, CTX_TOP);
  return {
    body,
    isNegated: (negationCount & 1) === 1,
    // Parser hint: a `**` was parsed inside `{}`; false proves the
    // lowering's separator-distribution check cannot trigger (§7).
    maybeSepDistribution: state.maybe_sep_distribution,
  };
}

function parseSequence(state, ctx) {
  const { input } = state;
  const nodes = [];
  let litBuf = [];

  function flushLit() {
    if (litBuf.length > 0) {
      nodes.push(lit(Uint8Array.from(litBuf)));
      litBuf = [];
    }
  }

  const inBrace = ctx.brace;
  while (state.pos < input.length) {
    const b = input[state.pos];

    // Brace context stops at the branch separator (`,`) or closer (`}`).
    if (inBrace && (b === COMMA || b === RBRACE)) break;

    switch (b) {
      case BACKSLASH: {
        // `\X` → literal X (lenient escape, GLOB_SPEC §9.1).
        state.pos++;
        if (state.pos >= input.length) throw new GlobError("TrailingBackslash");
        litBuf.push(input[state.pos]);
        state.pos++;
        break;
      }
      case SLASH:
        flushLit();
        nodes.push(sep());
        state.pos++;
        break;
      case QUESTION:
        flushLit();
        nodes.push(anyChar());
        state.pos++;
        break;
      case STAR:
        flushLit();
        parseStar(state, nodes, ctx);
        break;
      case LBRACK:
        flushLit();
        nodes.push(parseClass(state));
        break;
      case LBRACE: {
        flushLit();
        // Expanded-form neighbors for branch-edge globstar ownership
        // (§7 expansion equation / §8.1).
        const prevBoundary = boundaryBefore(nodes, ctx);
        const nextBoundary = braceNextBoundary(state, ctx);
        parseBraceInto(state, nodes, prevBoundary, nextBoundary);
        break;
      }
      default:
        // Anything else (including `@ ! ( ) |` and stray `] }`) is literal —
        // closers are meta only when paired with their opener (§9.1).
        litBuf.push(b);
        state.pos++;
    }
  }

  flushLit();

  // Single-child sequences elide the Concat wrapper.
  if (nodes.length === 1) return nodes[0];
  return concat(nodes);
}

function parseStar(state, nodes, ctx) {
  const { input } = state;
  // `**` is a globstar only when both sides are segment boundaries in
  // the EXPANDED form (§8.1, §7 equation) — see boundaryBefore/After.
  if (
    input[state.pos + 1] === STAR &&
    boundaryBefore(nodes, ctx) &&
    boundaryAfter(input[state.pos + 2], ctx)
  ) {
    if (state.brace_depth > 0) state.maybe_sep_distribution = true;
    nodes.push(globstar());
    state.pos += 2;
    // Collapse `/**/**` runs to a single globstar.
    while (
      state.pos + 3 <= input.length &&
      input[state.pos] === SLASH &&
      input[state.pos + 1] === STAR &&
      input[state.pos + 2] === STAR &&
      (state.pos + 3 === input.length || input[state.pos + 3] === SLASH)
    ) {
      state.pos += 3;
    }
    return;
  }

  // Single `*`, or degenerate `**` mid-segment — the second `*` is
  // consumed next iteration and folds into one Star.
  nodes.push(star());
  state.pos++;
}

function parseClass(state) {
  const { input } = state;
  const startPos = state.pos;
  state.pos++; // consume '['

  let negated = false;
  if (input[state.pos] === BANG || input[state.pos] === CARET) {
    negated = true;
    state.pos++;
  }

  const items = [];
  // POSIX: a leading `]` (after `[` or `[!`/`[^`) is a literal `]`.
  if (input[state.pos] === RBRACK) {
    items.push(classItemByte(RBRACK));
    state.pos++;
  }

  while (true) {
    if (state.pos >= input.length) throw new GlobError("UnterminatedClass", { at: startPos });
    const b = input[state.pos];
    if (b === RBRACK) {
      state.pos++;
      return klass(negated, items);
    }
    // Raw `/` mid-class means the class never closed inside its segment
    // (§6.2). Same outcome whether reached by raw or `\/`.
    if (b === SLASH) throw new GlobError("UnterminatedClass", { at: startPos });

    const lo = parseClassByte(state, startPos);
    if (
      input[state.pos] === DASH &&
      state.pos + 1 < input.length &&
      input[state.pos + 1] !== RBRACK
    ) {
      state.pos++; // consume '-'
      const hi = parseClassByte(state, startPos);
      if (hi < lo) throw new GlobError("InvalidRange", { at: startPos, low: lo, high: hi });
      items.push(classItemRange(lo, hi));
    } else {
      items.push(classItemByte(lo));
    }
  }
}

function parseClassByte(state, classStart) {
  const { input } = state;
  if (state.pos >= input.length) throw new GlobError("UnterminatedClass", { at: classStart });
  const b = input[state.pos];
  let resolved;
  if (b === BACKSLASH) {
    state.pos++;
    if (state.pos >= input.length) throw new GlobError("TrailingBackslash");
    resolved = input[state.pos];
    state.pos++;
  } else {
    resolved = b;
    state.pos++;
  }
  if (resolved === SLASH) throw new GlobError("UnterminatedClass", { at: classStart });
  return resolved;
}

// Append the parsed brace's nodes onto `nodes`. Single-branch braces
// `{a}` revert to literal `{a}` (GLOB_SPEC §7.4 — matches picomatch /
// fast-glob / bash).
function parseBraceInto(state, nodes, prevBoundary, nextBoundary) {
  const branches = parseBrace(state, prevBoundary, nextBoundary);
  if (branches.length === 1) {
    nodes.push(lit(Uint8Array.from([LBRACE])));
    const single = branches[0];
    const litBytes = nodeToLiteralBytes(single);
    if (litBytes !== null) {
      nodes.push(lit(litBytes));
    } else {
      nodes.push(single);
    }
    nodes.push(lit(Uint8Array.from([RBRACE])));
  } else {
    nodes.push(brace(branches));
  }
}

// Structural class skip: `[!`/`[^`, POSIX first-`]` literal, `\` escapes;
// stops after `]` or at `/` (which the real parser rejects later anyway).
function skipClassCandidate(input, start) {
  let i = start;
  if (input[i] === BANG || input[i] === CARET) i++;
  if (input[i] === RBRACK) i++;
  while (i < input.length && input[i] !== RBRACK && input[i] !== SLASH) {
    if (input[i] === BACKSLASH) i++;
    i++;
  }
  return Math.min(i + 1, input.length);
}

// Scan ahead from the current `{` to its matching `}` (honoring
// escapes, class scopes, and nesting) and report whether the byte
// after it is a boundary in the expanded form. Read-only — malformed
// tails error out in the real parse, so their value is a don't-care.
function braceNextBoundary(state, ctx) {
  const { input } = state;
  let i = state.pos + 1;
  let depth = 0;
  while (i < input.length) {
    const b = input[i];
    if (b === BACKSLASH) {
      i = Math.min(i + 2, input.length);
    } else if (b === LBRACK) {
      i = skipClassCandidate(input, i + 1);
    } else if (b === LBRACE) {
      depth++;
      i++;
    } else if (b === RBRACE) {
      if (depth === 0) return boundaryAfter(input[i + 1], ctx);
      depth--;
      i++;
    } else {
      i++;
    }
  }
  return true; // unterminated — errors in the real parse
}

function parseBrace(state, prevBoundary, nextBoundary) {
  const { input } = state;
  const startPos = state.pos;
  state.pos++; // consume '{'
  state.brace_depth++;
  if (state.brace_depth > MAX_BRACE_NESTING) {
    throw new GlobError("BraceNestingTooDeep", { max: MAX_BRACE_NESTING });
  }
  const ctx = { brace: true, prevBoundary, nextBoundary };
  const branches = [];
  while (true) {
    branches.push(parseSequence(state, ctx));
    const next = input[state.pos];
    if (next === COMMA) {
      state.pos++;
      continue;
    }
    if (next === RBRACE) {
      state.pos++;
      state.brace_depth--;
      return branches;
    }
    throw new GlobError("UnterminatedBrace", { at: startPos });
  }
}

File: crates/globstar/src/parser.rs
use crate::ast::*;
use crate::error::*;

pub fn parse(input: &[u8]) -> Result<Ast, GlobError> {
    if input.is_empty() {
        return Err(GlobError::Empty);
    }
    if input.len() > MAX_PATTERN_LEN {
        return Err(GlobError::TooLong {
            len: input.len(),
            max: MAX_PATTERN_LEN,
        });
    }

    let mut p = Parser {
        input,
        pos: 0,
        brace_depth: 0,
    };

    let mut negation_count = 0u32;
    while p.pos < input.len() && input[p.pos] == b'!' {
        negation_count += 1;
        p.pos += 1;
    }

    let body = p.parse_sequence(SequenceContext::Top)?;
    Ok(Ast {
        negation_count,
        body,
    })
}

#[derive(Clone, Copy)]
enum SequenceContext {
    /// Top-level: stop at end of input.
    Top,
    /// Inside `{...}`: stop at `,` or `}` at the matching depth.
    Brace {
        /// Expanded-form neighbor before the `{`: pattern start or a
        /// separator (chained through nested braces).
        prev_boundary: bool,
        /// Expanded-form neighbor after the matching `}`.
        next_boundary: bool,
    },
}

impl SequenceContext {
    fn boundary_before(self, last: Option<&Node>) -> bool {
        match last {
            None => matches!(
                self,
                Self::Top
                    | Self::Brace {
                        prev_boundary: true,
                        ..
                    }
            ),
            Some(Node::Separator) => true,
            Some(node @ Node::Brace(_)) => node_trails_boundary(node),
            _ => false,
        }
    }

    fn boundary_after(self, next: Option<u8>) -> bool {
        match next {
            None | Some(b'/') => true,
            Some(b',') | Some(b'}') => {
                matches!(
                    self,
                    Self::Brace {
                        next_boundary: true,
                        ..
                    }
                )
            }
            _ => false,
        }
    }
}

struct Parser<'a> {
    input: &'a [u8],
    pos: usize,
    brace_depth: usize,
}

impl<'a> Parser<'a> {
    fn peek(&self) -> Option<u8> {
        self.input.get(self.pos).copied()
    }

    fn peek_at(&self, offset: usize) -> Option<u8> {
        self.input.get(self.pos + offset).copied()
    }

    fn parse_sequence(&mut self, ctx: SequenceContext) -> Result<Node, GlobError> {
        let remaining = self.input.len() - self.pos;
        let node_capacity = match ctx {
            SequenceContext::Top => remaining / 2 + 1,
            SequenceContext::Brace { .. } => (remaining / 2 + 1).min(8),
        };
        let mut nodes: Vec<Node> = Vec::with_capacity(node_capacity);
        let mut lit_buf: Vec<u8> = Vec::with_capacity(remaining.min(32));

        let in_brace = matches!(ctx, SequenceContext::Brace { .. });
        while self.pos < self.input.len() {
            let b = self.input[self.pos];
            if in_brace && (b == b',' || b == b'}') {
                break;
            }
            match b {
                b'\\' => {
                    self.pos += 1;
                    if self.pos >= self.input.len() {
                        return Err(GlobError::TrailingBackslash);
                    }
                    if self.input[self.pos] == b'/' {
                        return Err(GlobError::EscapedSeparator { at: self.pos - 1 });
                    }
                    lit_buf.push(self.input[self.pos]);
                    self.pos += 1;
                }
                b'/' => {
                    flush_literal(&mut lit_buf, &mut nodes);
                    nodes.push(Node::Separator);
                    self.pos += 1;
                }
                b'?' => {
                    flush_literal(&mut lit_buf, &mut nodes);
                    nodes.push(Node::AnyChar);
                    self.pos += 1;
                }
                b'*' => {
                    flush_literal(&mut lit_buf, &mut nodes);
                    self.parse_star(&mut nodes, ctx);
                }
                b'[' => {
                    flush_literal(&mut lit_buf, &mut nodes);
                    let class = self.parse_class()?;
                    nodes.push(Node::Class(class));
                }
                b'{' => {
                    let (single, next_after_brace) = self.scan_brace(ctx);
                    let (prev_boundary, next_boundary) = if single {
                        (false, false)
                    } else {
                        (
                            lit_buf.is_empty() && ctx.boundary_before(nodes.last()),
                            next_after_brace,
                        )
                    };
                    match <[Node; 1]>::try_from(self.parse_brace(prev_boundary, next_boundary)?) {
                        Ok([single]) => {
                            lit_buf.push(b'{');
                            match single.to_literal_bytes() {
                                Some(bytes) if !bytes.contains(&b'/') => {
                                    lit_buf.extend_from_slice(&bytes);
                                }
                                _ => {
                                    flush_literal(&mut lit_buf, &mut nodes);
                                    nodes.push(single);
                                }
                            }
                            lit_buf.push(b'}');
                        }
                        Err(branches) => {
                            flush_literal(&mut lit_buf, &mut nodes);
                            nodes.push(Node::Brace(branches));
                        }
                    }
                }
                _ => {
                    lit_buf.push(b);
                    self.pos += 1;
                }
            }
        }

        flush_literal(&mut lit_buf, &mut nodes);

        Ok(match <[Node; 1]>::try_from(nodes) {
            Ok([single]) => single,
            Err(nodes) => Node::Concat(nodes),
        })
    }

    fn parse_star(&mut self, nodes: &mut Vec<Node>, ctx: SequenceContext) {
        debug_assert_eq!(self.input[self.pos], b'*');

        if self.peek_at(1) == Some(b'*')
            && ctx.boundary_before(nodes.last())
            && ctx.boundary_after(self.peek_at(2))
        {
            nodes.push(Node::Globstar);
            self.pos += 2;
            while self.pos + 3 <= self.input.len()
                && &self.input[self.pos..self.pos + 3] == b"/**"
                && (self.pos + 3 == self.input.len() || self.input[self.pos + 3] == b'/')
            {
                self.pos += 3;
            }
            return;
        }

        nodes.push(Node::Star);
        self.pos += 1;
    }

    fn parse_class(&mut self) -> Result<CharClass, GlobError> {
        let start_pos = self.pos;
        debug_assert_eq!(self.input[self.pos], b'[');
        self.pos += 1;

        let negated = matches!(self.peek(), Some(b'!') | Some(b'^'));
        if negated {
            self.pos += 1;
        }

        let mut items: Vec<ClassItem> = Vec::with_capacity(4);

        if self.peek() == Some(b']') {
            items.push(ClassItem::Byte(b']'));
            self.pos += 1;
        }

        loop {
            let b = match self.peek() {
                Some(b) => b,
                None => return Err(GlobError::UnterminatedClass { at: start_pos }),
            };
            if b == b']' {
                self.pos += 1;
                return Ok(CharClass { negated, items });
            }
            let low = self.parse_class_byte(start_pos)?;
            if self.peek() == Some(b'-') && self.peek_at(1) != Some(b']') {
                self.pos += 1; // consume `-`
                let high = self.parse_class_byte(start_pos)?;
                if high < low {
                    return Err(GlobError::InvalidRange {
                        at: start_pos,
                        low,
                        high,
                    });
                }
                items.push(ClassItem::Range(low, high));
            } else {
                items.push(ClassItem::Byte(low));
            }
        }
    }

    fn parse_class_byte(&mut self, class_start: usize) -> Result<u8, GlobError> {
        let b = self
            .peek()
            .ok_or(GlobError::UnterminatedClass { at: class_start })?;
        let resolved = if b == b'\\' {
            self.pos += 1;
            let next = self.peek().ok_or(GlobError::TrailingBackslash)?;
            self.pos += 1;
            next
        } else {
            self.pos += 1;
            b
        };
        if resolved == b'/' {
            return Err(GlobError::UnterminatedClass { at: class_start });
        }
        Ok(resolved)
    }

    fn scan_brace(&self, ctx: SequenceContext) -> (bool, bool) {
        debug_assert_eq!(self.input[self.pos], b'{');
        let input = self.input;
        let mut i = self.pos + 1;
        let mut depth = 0usize;
        let mut single = true;
        while i < input.len() {
            match input[i] {
                b'\\' => i = (i + 2).min(input.len()),
                b'[' => {
                    i += 1;
                    if matches!(input.get(i), Some(b'!') | Some(b'^')) {
                        i += 1;
                    }
                    if input.get(i) == Some(&b']') {
                        i += 1;
                    }
                    while i < input.len() && input[i] != b']' && input[i] != b'/' {
                        if input[i] == b'\\' {
                            i += 1;
                        }
                        i += 1;
                    }
                    i = (i + 1).min(input.len());
                }
                b'{' => {
                    depth += 1;
                    i += 1;
                }
                b',' if depth == 0 => {
                    single = false;
                    i += 1;
                }
                b'}' => {
                    if depth == 0 {
                        return (single, ctx.boundary_after(input.get(i + 1).copied()));
                    }
                    depth -= 1;
                    i += 1;
                }
                _ => i += 1,
            }
        }
        (single, true)
    }

    fn parse_brace(
        &mut self,
        prev_boundary: bool,
        next_boundary: bool,
    ) -> Result<Vec<Node>, GlobError> {
        let start_pos = self.pos;
        debug_assert_eq!(self.input[self.pos], b'{');
        self.pos += 1;

        self.brace_depth += 1;
        if self.brace_depth > MAX_BRACE_NESTING {
            return Err(GlobError::BraceNestingTooDeep {
                max: MAX_BRACE_NESTING,
            });
        }

        let mut branches = Vec::with_capacity(4);
        loop {
            let branch = self.parse_sequence(SequenceContext::Brace {
                prev_boundary,
                next_boundary,
            })?;
            branches.push(branch);

            match self.peek() {
                Some(b',') => {
                    self.pos += 1;
                    continue;
                }
                Some(b'}') => {
                    self.pos += 1;
                    self.brace_depth -= 1;
                    return Ok(branches);
                }
                _ => return Err(GlobError::UnterminatedBrace { at: start_pos }),
            }
        }
    }
}

fn flush_literal(buf: &mut Vec<u8>, nodes: &mut Vec<Node>) {
    if !buf.is_empty() {
        nodes.push(Node::Literal(std::mem::take(buf)));
    }
}

fn node_trails_boundary(node: &Node) -> bool {
    match node {
        Node::Separator => true,
        Node::Concat(children) => children.last().is_some_and(node_trails_boundary),
        Node::Brace(branches) => !branches.is_empty() && branches.iter().all(node_trails_boundary),
        _ => false,
    }
}
